eventuate-adapter-vertx/src/main/scala/com/rbmhtechnology/eventuate/adapter/vertx/VertxReader.scala

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2015 - 2016 Red Bull Media House GmbH <http://www.redbullmediahouse.com> - all rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.rbmhtechnology.eventuate.adapter.vertx
+
+import com.rbmhtechnology.eventuate.EventsourcedWriter
+import com.rbmhtechnology.eventuate.adapter.vertx.api.{ StorageProvider, VertxEndpointRouter }
+import io.vertx.core.eventbus.impl.MessageImpl
+import io.vertx.core.eventbus.{ DeliveryOptions, EventBus, Message }
+import io.vertx.core.{ AsyncResult, Handler, Vertx, Future => VertxFuture }
+
+import scala.collection.immutable.Seq
+import scala.concurrent.duration.{ FiniteDuration, _ }
+import scala.concurrent.{ ExecutionContext, Future, Promise }
+
+case class EventEnvelope(address: String, event: Any)
+
+trait MessageProducer {
+  import VertxHandlerConverters._
+
+  protected def eventBus(): EventBus =
+    vertx.eventBus()
+
+  def vertx: Vertx
+
+  protected def produce[A](address: String, msg: Any, deliveryOptions: DeliveryOptions, handler: Handler[AsyncResult[Message[A]]]): Unit
+
+  def produce(address: String, msg: Any): Unit = {
+    produce[Unit](address, msg, new DeliveryOptions(), ((ar: AsyncResult[Message[Unit]]) => {}).asVertxHandler)
+  }
+
+  def produce[A](address: String, msg: Any, timeout: FiniteDuration = 30.seconds)(implicit ec: ExecutionContext): Future[A] = {
+    val promise = Promise[Message[A]]
+    produce(address, msg, new DeliveryOptions().setSendTimeout(timeout.toMillis), promise.asVertxHandler)
+    promise.future.map(_.body)
+  }
+}
+
+trait MessagePublisher extends MessageProducer {
+  override protected def produce[A](address: String, msg: Any, deliveryOptions: DeliveryOptions, handler: Handler[AsyncResult[Message[A]]]): Unit = {
+    eventBus().publish(address, msg, deliveryOptions)
+    handler.handle(VertxFuture.succeededFuture(new MessageImpl[Any, A]()))
+  }
```

Comment:
In this case you may want to abstract over message delivery somewhere else (e.g. on reader-level) rather than here. Completing an asynchronous handler where the return from a from a method call is sufficient is a code/design smell IMO.